server/src/main/java/org/apache/druid/server/coordinator/ClientCompactionRunnerInfo.java

```diff
@@ -0,0 +1,198 @@
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    ClientCompactionRunnerInfo that = (ClientCompactionRunnerInfo) o;
+    return type == that.type;
+  }
+
+  public static class ValidationResult{
+    private final boolean valid;
+    private final String reason;
+
+    public ValidationResult(boolean valid, String reason)
+    {
+      this.valid = valid;
+      this.reason = reason;
+    }
+
+    public boolean isValid()
+    {
+      return valid;
+    }
+
+    public String getReason()
+    {
+      return reason;
+    }
+  }
+
+  @Override
+  public int hashCode()
+  {
+    return Objects.hash(type);
+  }
+
+  public static ValidationResult validateCompactionConfig(
+      DataSourceCompactionConfig newConfig,
+      CompactionEngine defaultCompactionEngine
+  )
+  {
+    CompactionEngine compactionEngine = newConfig.getEngine() == null ? defaultCompactionEngine : newConfig.getEngine();
+    if (compactionEngine == CompactionEngine.NATIVE) {
+      return new ValidationResult(true, null);
+    } else {
+      return msqEngineSupportsCompactionConfig(newConfig);
+    }
+  }
+
+  /**
+   * Checks if the provided compaction config is supported by MSQ. The following configs aren't supported:
+   * <ul>
+   * <li>finalizeAggregations set to false in context.</li>
+   *
+   * <li>partitionsSpec of type HashedParititionsSpec.</li>
+   *
+   * <li>maxTotalRows in DynamicPartitionsSpec.</li>
+   *
+   * <li>rollup set to false in granularitySpec when metricsSpec is specified.</li>
+   * </ul>
+   *
+   * @param newConfig The updated compaction config
+   * @return ValidationResult. The reason string is null if isValid() is True.
+   */
+  private static ValidationResult msqEngineSupportsCompactionConfig(DataSourceCompactionConfig newConfig)
+  {
+    if (newConfig.getTuningConfig() != null) {
+      ValidationResult partitionSpecValidationResult =
+          validatePartitionsSpec(newConfig.getTuningConfig().getPartitionsSpec());
+      if (!partitionSpecValidationResult.isValid()) {
+        return partitionSpecValidationResult;
+      }
+    }
+    if (newConfig.getGranularitySpec() != null) {
+      ValidationResult rollupValidationResult = validateRollup(
+          newConfig.getMetricsSpec(),
+          newConfig.getGranularitySpec().isRollup()
+      );
+      if (!rollupValidationResult.isValid()) {
+        return rollupValidationResult;
+      }
+    }
+    return new ValidationResult(true, null);
+  }
+
+  /**
+   * Validte that partitionSpec is either 'dynamic` or 'range', and if 'dynamic', ensure maxTotalRows is null.
+   */
+  public static ValidationResult validatePartitionsSpec(PartitionsSpec partitionsSpec)
+  {
+    if (!(partitionsSpec instanceof DimensionRangePartitionsSpec
+          || partitionsSpec instanceof DynamicPartitionsSpec)) {
+      return new ValidationResult(false, StringUtils.format(
```

Comment:
Accept `Object.. args` in the constructor of `ValidationResult` and do the `StringUtils.format` in the constructor itself.